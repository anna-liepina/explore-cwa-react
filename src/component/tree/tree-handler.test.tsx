import React from 'react';
import '@testing-library/jest-dom';
import { render, fireEvent } from '@testing-library/react';
import TreeHandler from './tree-handler';

describe('<TreeHandler/>', () => {
    const data = [
        {
            text: 'root node',
            nodes: [
                {
                    text: 'child node 1 - with children',
                    nodes: [
                        {
                            text: 'inner child node',
                        }
                    ]
        
                },
                {
                    text: 'child node 2 - without children',
                }
            ]
        },
    ];

    const props = {
        onFilter: jest.fn(),
        onExpand: jest.fn(),
        data,
    };

    const optionalProps = {
        'data-cy': 'optProps.data-cy',
        className: 'optProps.className',
        title: 'optProps.title',
        patternPlaceholder: 'optProps.patternPlaceholder'
    };

    describe('render', () => {
        it('with default/required props', () => {
            const { asFragment } = render(<TreeHandler {...props} />);

            expect(asFragment()).toMatchSnapshot();
        });

        it('with optional/required props', () => {
            const { asFragment } = render(<TreeHandler {...props} {...optionalProps} />);

            expect(asFragment()).toMatchSnapshot();
        });

        it('if node has ::isVisible = false, it should NOT render it', () => {
            const { asFragment } = render(<TreeHandler {...props} data={data.map((v) => ({ ...v, isVisible: false }))} />);

            expect(asFragment()).toMatchSnapshot();
        });
    });

    describe('internal callbacks', () => {
        /** functionality relay on data-node attribute */
        describe('::onExpand', () => {
            it('should invoke external callback [::onExpand] with [::data] state field and value of ["data-node"] from on a click on [data-cy="tree-node-0"]', () => {
                const spy = jest.fn();
                const { container } = render(<TreeHandler {...props} data={data} onExpand={spy} />);

                fireEvent.click(container.querySelector('[data-cy="tree-node-0"]')!);

                expect(spy).toBeCalledWith(
                    data,
                    "0"
                );
            });

            it('should invoke external callback [::onExpand] with path from ["data-node"] and [::data] state field', () => {
                const spy = jest.fn();
                const { container } = render(<TreeHandler {...props} data={data} onExpand={spy} />);

                fireEvent.click(container.querySelector('section')!);

                expect(spy).not.toBeCalled();
            });
        });

        /** functionality relay on data-section and data-field attributes */
        describe('::onChange', () => {
            it('should invoke external callback [::onFilter] with relevant payload from a change event of [data-cy="tree-pattern"]', () => {
                const spy = jest.fn();
                const { container } = render(<TreeHandler {...props} data={data} onFilter={spy} />);

                fireEvent.change(container.querySelector('[data-cy="tree-pattern"]')!, { target: { value: 'val' } });

                expect(spy).toBeCalledWith(
                    data,
                    "val"
                );
            });
        });
    });
});
