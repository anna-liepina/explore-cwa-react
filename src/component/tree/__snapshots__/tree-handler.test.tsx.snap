// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeHandler/> render if node has ::isVisible = false, it should NOT render it 1`] = `
<DocumentFragment>
  <section
    class="tree"
  >
    <input
      class="tree_input"
      data-cy="tree-pattern"
      value=""
    />
    <div
      class="tree_content"
    />
  </section>
</DocumentFragment>
`;

exports[`<TreeHandler/> render with default/required props 1`] = `
<DocumentFragment>
  <section
    class="tree"
  >
    <input
      class="tree_input"
      data-cy="tree-pattern"
      value=""
    />
    <div
      class="tree_content"
    >
      <div
        class="tree-node--with-children"
        data-cy="tree-node-0"
        data-node="0"
      >
        <span>
          root node
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`<TreeHandler/> render with optional/required props 1`] = `
<DocumentFragment>
  <section
    class="tree optProps.className"
  >
    <input
      class="tree_input"
      data-cy="optProps.data-cytree-pattern"
      placeholder="optProps.patternPlaceholder"
      value=""
    />
    <div
      class="tree_content"
    >
      <div
        class="tree-node--with-children"
        data-cy="optProps.data-cytree-node-0"
        data-node="0"
      >
        <span>
          root node
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
